### PR TITLE
Adds FIDO U2F support

### PR DESCRIPTION
**Pull request by konklone**
This adds two factor authentication to my blog's admin area, using [the open U2F standard](https://www.yubico.com/applications/fido/), developed by the [FIDO Alliance](https://fidoalliance.org/), and [now used by Google](http://googleonlinesecurity.blogspot.com/2014/10/strengthening-2-step-verification-with.html) to secure Google accounts.

[fido-key-1](https://konklone.com/post/get-a-fido-key-right-now-and-log-into-stuff-with-it)

**For now**, this only works in Chrome 39+, and depends on the user having the [FIDO U2F Chrome extension](https://chrome.google.com/webstore/detail/fido-u2f-universal-2nd-fa/pfboblefjcgdjicmnffhdgionmgcdmne) installed. Future versions of Chrome will remove the extension requirement. (Google is currently accomplishing this for their accounts in Chrome without requiring an extension by literally whitelisting their own domains in their browser.)

My blog is written in Ruby, using the [Sinatra microframework](http://www.sinatrarb.com/). It saves device data into a [MongoDB](https://www.mongodb.org/) database, using [Mongoid](http://mongoid.org). I I used the [`ruby-u2f`](https://github.com/userbin/ruby-u2f) library ([linked to](https://developers.yubico.com/U2F/Libraries/List_of_libraries.html) by Yubico), and an [$18 FIDO U2F Security Key](http://www.amazon.com/gp/product/B00NLKA0D8/ref=oh_aui_detailpage_o01_s00?ie=UTF8&psc=1) made by Yubico.

**This example uses no user account system**, as I'm the only user on the system. The admin password is stored in the app's global configuration, and there is one database table that stores all the devices I register. This makes the example simple, but systems with user accounts may have important differences (like validating that the device used for login is registered to the specific user it was registered for).

## Workflow

While **logged in**, to register a new device, I visit the new **U2F** area of the admin section.

01-no-devices

To register a new device, I type in a name I want to give it, and hit the "Register Key" button. This causes the FIDO U2F Chrome extension to pop up an Allow/Block dialog. (This will be rendered moot in future versions of Chrome.)

02-approve-block

If I choose **Block**, it generates an error message about not being able to reach the USB device.

03-bad-device

If I do this again and instead choose **Allow**, and then **hit the button**, it registers the device and assigns it the name I gave.

04-registered

I can show the details of the four U2F-required fields associated with the device: the key handle (handy unique ID for the database), a certificate, a Base64-encoded public key, and a counter whose value will increase with every registration and login attempt (used to prevent replay attacks and detect device cloning). 

05-with-details

(I'm not sure of which, if any, of these are sensitive to reveal, but they are all uniquely generated for each registration, and were only used for my laptop's local instance.)

If I try to register the same key again, I can detect an error.

06-already-registered

Now if I **log out**, and then **log in**, first with my normal password:

07-login-1

This looks up whether or not I have a device registered, and if so, only "half" logs me in, and directs me to the key authentication step.

08-login-2

At this point I can just **hit the button** and log in.

09-logged-in

Next, the implementation details.

## Device registration

My app has an `Environment.u2f` object globally available, initialized with a U2F APP_ID that is the site's root, e.g. `http://localhost:8080` or `https://konklone.com`.

```ruby
class Environment
  def self.u2f
    @u2f ||= U2F::U2F.new(Environment.config['site']['root'])
  end
end
```

The key/registration page endpoint:

* loads a "registration request" for every version of U2F a device might plausibly support
* stores each request's associated challenge in the session
* looks up any of my already activated devices so that the registration process can detect already registered devices

```ruby
get '/admin/key/register' do
  # Generate one for each version of U2F, currently only `U2F_V2`
  registration_requests = Environment.u2f.registration_requests

  # Keep challenges around for verification
  session[:challenges] = registration_requests.map &:challenge

  # Key handles for all registered devices (to me: which is all of them)
  devices = Device.all
  key_handles = devices.map &:key_handle
  sign_requests = Environment.u2f.authentication_requests key_handles

  erb :"admin/key_register", layout: :"admin/layout", locals: {
    registration_requests: registration_requests,
    sign_requests: sign_requests,
    devices: devices
  }
end
```

This data is sent to the template, and used to create a form and set up some JavaScript that will run when the Register Key button is pressed:

01-no-devices

```html
<!-- needed to do u2f work below -->
<script src="chrome-extension://pfboblefjcgdjicmnffhdgionmgcdmne/u2f-api.js"></script>

<button disabled onclick="startRegister()">Register a new device:</button>

<form id="register" action="/admin/key/register" method="post">
  <input type="hidden" name="response" />
  <input type="text" placeholder="device name" required name="name" />
</form>
```

```javascript
var codes = {
  "1": "Other, unknown error.",
  "2": "Bad request, or blocked USB access.",
  "3": "Configuration unsupported.",
  "4": "Device already registered, or ineligible.",
  "5": "Timeout when waiting for device interaction."
};

var startRegister = function() {
  if ($.trim($(".register input[name=name]").val()) == "")
    return log("You gotta name the device.");

  u2f.register(registerRequests, signRequests, function(registerResponse) {
    if (registerResponse.errorCode) {
      log("Registration error: " + codes[registerResponse.errorCode]);
      return console.log("Registration error: " + codes[registerResponse.errorCode]);
    }

    var form = document.getElementById("register");
    var response = document.querySelector('[name=response]');
    response.value = JSON.stringify(registerResponse);

    form.submit();
  });
};
```

The `u2f.register` method kicks off the Allow/Block dialogue. When Allow is pressed, the registerRequests are sent to the token. When the token's button is pressed, the callback fires and returns an error code and/or an object with response details from the token. If there's no error code, the response details are POSTed to the server.

The server receives the registration response, validates that it has a name, and:

* Loads and validates the response from the token, emailing the admin and warning the user if the token sent up invalid data.
* Attempts to register the token, which actually compares the challenge to the response. If this fails (e.g. there was a malicious actor in the middle or something) the device registration fails.
* If device registration succeeds, the device's registration data is saved in the database.

```ruby
post '/admin/key/register' do
  unless (name = params[:name]).present?
    flash[:failure] = "I need a device name."
    redirect "/admin/key/register"
  end

  begin
    response = U2F::RegisterResponse.load_from_json params[:response]
  rescue Exception => exc
    Email.exception exc, {response: params[:response]}
    flash[:failure] = "Invalid registration data."
    redirect "/admin/key/register"
  end

  reg = begin
    Environment.u2f.register!(session[:challenges], response)
  rescue U2F::Error => exc
    Email.exception exc
    nil
  ensure
    session.delete :challenges
  end

  if reg
    flash[:success] = "DEVICE REGISTERED. THANK YOU, TOKEN BEARER."

    Device.create!(
      certificate: reg.certificate,
      key_handle:  reg.key_handle,
      public_key:  reg.public_key,
      counter:     reg.counter,
      name: name
    )
  else
    flash[:failure] = "DEVICE NOT REGISTERED. EMAIL SENT WITH YOUR FAILURE."
  end

  redirect "/admin/key/register"
end
```

04-registered

## Device login

07-login-1

After the user logs in with a normal password, the app checks if the user has a device registered. If so, the user is not fully logged in, and is instead redirected to the key login step.

```ruby
# This is after the user has *submitted* their normal password.
post '/admin/login' do
  if params[:password] == Environment.config['admin']['password']

    if Device.count > 0
      session[:half_admin] = true
      redirect '/admin/key/login'
    else
      session[:admin] = true
      redirect '/admin/posts/published'
    end
  else
    erb :"admin/login", layout: :"admin/layout", locals: {message: "Invalid credentials."}
  end
end
```

The device login page looks up any registered devices, creates "signing requests" for each of them, and stores each request's associated challenge in the session.

```ruby
get '/admin/key/login' do
  unless half_admin?
    flash[:failure] = "You need to know the password before you get to show the token."
    redirect "/admin"
  end

  key_handles = Device.all.map &:key_handle

  if key_handles.empty?
    flash[:failure] = "Weird: no keys registered. Why are you here?"
    redirect "/admin"
  end

  sign_requests = Environment.u2f.authentication_requests key_handles

  session[:challenges] = sign_requests.map &:challenge

  erb :"admin/key_login", layout: :"admin/layout", locals: {
    sign_requests: sign_requests
  }
end
```

The signing requests are sent to the template, which uses a hidden form to collect and post the response once submitted. Some JavaScript uses the U2F API to send the login request:

08-login-2

```html
<!-- needed to do u2f work below -->
<script src="chrome-extension://pfboblefjcgdjicmnffhdgionmgcdmne/u2f-api.js"></script>

<section class="login">
  Log in with your key now by hitting the button.
</section>

<form id="authenticate" action="/admin/key/login" method="post">
  <input type="hidden" name="response" />
</form>
```

```javascript
var signRequests = <%= sign_requests.as_json.to_json.html_safe %>;

u2f.sign(signRequests, function(signResponse) {
  if (signResponse.errorCode)
    return console.log("Authentication error: " + codes[signResponse.errorCode]);

  console.log("Authentication message received.")
  console.log(signResponse);

  var form = document.getElementById("authenticate");
  var response = document.querySelector('[name=response]');
  response.value = JSON.stringify(signResponse);

  form.submit();
});
```

The `u2f.sign` call prepares the page to listen for when the device's button is pressed. When the button is pressed, the response is read in by the page and the callback fires with a response code and/or some response details. The response details are then inserted into the hidden form, and the form is submitted.

Finally, the endpoint that receives the device login information:

* Verifies the user is half-logged in already.
* Loads the response data the token provided.
* Looks up the device by the provided handle to ensure it's been registered.
* Authenticates the device based on the server's challenge, the token's response, the original registered device's public certificate, and the last known counter value for the registered device.
* If the given counter was too low, this fails the login *and* suggests that the user's device may have been tampered with or cloned. A future version of my implementation will deactivate the device automatically if it sees this condition.
* If there was some other authentication error, this fails the login.
* If the login was successful, update the device's record with the new counter value, and log the user in.

```ruby
post '/admin/key/login' do
  unless half_admin?
    flash[:failure] = "You need to know the password before you get to show the token."
    redirect "/admin"
  end

  response = U2F::SignResponse.load_from_json params[:response]

  unless device = Device.where(key_handle: response.key_handle).first
    flash[:failure] = "This device has never been registered."
    redirect "/admin/key/login"
  end

  authenticated = false
  failure = nil
  begin
    Environment.u2f.authenticate!(
      session[:challenges],
      response,

      # database stores base64-encoded version - library needs real binary string
      Base64.strict_decode64(device.public_key),

      device.counter
    )
    authenticated = true

  rescue U2F::CounterToLowError => exc
    Email.exception exc
    failure = "Your device has gotten out of sync with the server. Your device may have been compromised. You will need to use other means to invalidate this device and re-authorize it or add a new one."
    nil
  rescue U2F::Error => exc
    Email.exception exc
    failure = "Failed to log in. Try again or something?"
    nil
  ensure
    session.delete :challenges
  end

  if authenticated
    device.update(counter: response.counter)
    flash[:success] = "WELCOME, TOKEN BEARER."

    session.delete :half_admin
    session[:admin] = true

    redirect "/admin"
  else
    flash[:failure] = failure
    redirect "/admin/key/login"
  end
end
```

09-logged-in

Phew! For all that code, the user experience is very simple, and pressing that button Just Works.

## Further work

* This is **missing unit tests**, which I had a hard time thinking of how to structure before I learned how this worked through implementing. That needs to be done next.
* If the `counter` value comes in too low on log in, this is bad: it suggests that the device has been cloned or in some other way tampered with. However, just deleting the device altogether removes the factor entirely, which would make the factor useless. Instead, I plan to have the app "deactivate" (but not deregister) a token, but still use the presence of deactivated tokens to insist on using a token to log in. This effectively locks me out until I can SSH into the server and reset things.

## Security model

The application's security model informs the implementation:

* I am comfortable implementing a U2F token system with no backup plan because I can always SSH into the server (which has a separate security model) and reset the database (or even alter the code) to let me temporarily log in without a token or backup code. Assuming I keep my SSH credentials safe, the presence of this side channel does not weaken the second factor's protection against remote entry attempts.
* In Google's implementation, they don't require a U2F token to log in -- device tokens are used as an *additional* option for a second factor. Because they are the strongest possible factor, Google presents you the device option first. If you don't have your device on you, you can use any of your other second factors (OTP, backup codes) instead. However, since Google lets you decline to use your key and use your OTP method instead, it's not clear to me how they'd protect the user against a MITM phisher who invisibly told Google to move on to OTP.
* In a team environment, where U2F tokens are used to guarantee access to internally shared resources, tokens could be required in a more absolute way (without backup codes for each teammate floating around out there), because you could always be "rescued" by a team admin. As long as you had multiple team admins who could rescue each other, the system would be relatively strong against lost tokens, at the cost of waiting on teammates to find the time to rescue you. The rescue process might consist of an ephemeral, short-lived backup password generated by the admin and given to the teammate over a secure channel.

## Resources

* [Yubico's enthusiastic explanation](https://www.yubico.com/applications/fido/) of the benefits of U2F.
* If you're on Linux, you need to [add a udev rule](https://konklone.com/post/get-a-fido-key-right-now-and-log-into-stuff-with-it#getting-it-working-on-linux). (_shakes fist_)
* [Discussion on Yubico's PHP U2F server implementation](https://github.com/Yubico/php-u2flib-server/issues/5) on how to handle counter failures, and the role of a library vs application.

### CODE REVIEW

**Comment by paultag** [comment]
:confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball:

**Comment by adelevie** [comment]
:dragon: :performing_arts: :shipit: :trumpet: :handbag: :turtle:

File: app/views/admin/post.erb [hunk 1]
```diff
@@ -159,29 +156,13 @@
...
-  $("button.preview").click(function() {
-    $("#preview .title").val($("input.title").val());
-    $("#preview .body").val($("textarea.body").val());
-    $("#preview .footer").val($("textarea.footer").val());
-    $("#preview").get(0).submit();
-    return false;
+      return false;
+    });
   });
-});
```
Comment (konklone) [inline app/views/admin/post.erb]: This removal of preview markup was unrelated to U2F work.

**Comment by konklone** [comment]
Some thoughts for later:

* How does U2F resist, if at all, downgrade attacks when U2F is just one of several 2FA options (e.g. Google)?
* How does U2F resist, if at all, an MITM with the ability to DNS poison a domain *and* produce a phony certificate the client will trust for that domain?
* I really need to dig into the [implementation considerations](http://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-u2f-implementation-considerations-ps-20141009.html) and [security reference](http://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-security-ref-ps-20141009.html) for U2F.

**Pull request merged**